cognite/client/data_classes/assets.py

```diff
@@ -449,3 +456,244 @@ def dump(self, camel_case: bool = False) -> Dict[str, Any]:
+    ) -> None:
+        if self.__validation_has_run:  # Validation is expensive
+            return
+
+        self._roots, self._orphans, self._missing, self._duplicates = self._inspect_attributes()
+        if verbose:
+            self._report_on_identifiers(output_file)
+
+        if self._preconditions_for_cycle_check_are_met(on_error):
+            self._cycle_subtree_size, self._cycles = self._locate_cycles(on_error=on_error)
+            if verbose:
+                self._report_on_cycles(output_file)
+        elif verbose:
+            self.print_to(
+                "\nUnable to check for cyclical references before above issues are fixed", output_file=output_file
+            )
+
+        self.__validation_has_run = True
+        self.is_valid(on_error=on_error)
+
+    def validate_and_report(self, output_file: Optional[Path] = None) -> None:
+        self.validate(verbose=True, output_file=output_file, on_error="ignore")
+
+    def groupby_parent_xid(self) -> Dict[Optional[str], List[Asset]]:
+        self.is_valid(on_error="raise")
+        # Sort (on parent) as required by groupby. Since the empty string "" is a valid external ID,
+        # we sort based on the hash of the parent - which may be None:
+        parent_sorted = sorted(self._assets, key=lambda a: hash(self._get_parent(a)))
+        return {
+            parent: list(child_assets)
+            for parent, child_assets in itertools.groupby(parent_sorted, key=self._get_parent)
+        }
+
+    def _inspect_attributes(self) -> Tuple[List[Asset], List[Asset], List[Asset], Dict[str, List[Asset]]]:
+        missing, orphans, roots, duplicates = [], [], [], defaultdict(list)
+        xid_count = Counter(map(self._get_xid, self._assets))
+
+        for asset in self._assets:
+            xid, parent, name = asset.external_id, asset.parent_external_id, asset.name
+            if xid is None or name is None or len(name) < 1:
+                missing.append(asset)
+                continue  # Don't report missing as part of any other group
+
+            if xid_count[xid] > 1:
+                duplicates[xid].append(asset)
+
+            if parent is None:
+                roots.append(asset)
+            elif parent not in xid_count:
+                orphans.append(asset)
+
+        return roots, orphans, missing, dict(duplicates)
+
+    def _preconditions_for_cycle_check_are_met(self, on_error: OnErrorOptions) -> bool:
+        # Required before cycle-check can run
+        if not (not self._roots or self._orphans or self._missing or self._duplicates):
+            return True
+        self._on_error(on_error, "Unable to run cycle-check before basic issues are fixed.")
+        return False
+
+    def _locate_cycles(self, on_error: OnErrorOptions) -> Tuple[int, List[List[str]]]:
+        edges = dict(map(op.attrgetter("external_id", "parent_external_id"), self._assets))
+        has_cycles, no_cycles = set(), {None, *map(self._get_xid, self.roots)}
```

Comment:
Not sure what no_cycles contains at its start.

Reply:
Do you want me to add a comment? It's just an optimization to add all root asset external IDs to the set of assets that are not part of a cycle. I also add `None` to avoid `if x in no_cycles or x is None` below because whenever a parent external ID is `None`, you are for sure safe 😜